From 7031c4391115946db0c18310c3a8dabf50ba3266 Thu Jul 19 09:25:01 2018 +0300
From: Zdenek Valek <zdenek.valek@nxp.com>
Date: Thu, 9 Aug 2018 17:16:00 +0200
Subject: [PATCH] vlan: add vlan for VSDK specific configuration

Signed-off-by: Ludovit Minarik <ludovit.minarik@nxp.com>
Signed-off-by: Zdenek Valek <zdenek.valek@nxp.com>
Signed-off-by: Catalin Udma <catalin-dan.udma@nxp.com>
Signed-off-by: Daniel Lazar <daniel.lazar@nxp.com>
---

diff --git a/tools/firmware_generation/SJA1105Q_firmware_sw0_bbmini.py b/tools/firmware_generation/SJA1105Q_firmware_sw0_bbmini.py
index 55b7435..007356e 100755
--- a/tools/firmware_generation/SJA1105Q_firmware_sw0_bbmini.py
+++ b/tools/firmware_generation/SJA1105Q_firmware_sw0_bbmini.py
@@ -234,6 +234,43 @@ vlan_lookup_table.append({
     "TAG_PORT"   : 0x00, # Egress frames are untagged
     "VLANID"     : default_vlan})
 
+# Camera VLANs
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x11, # Ports 4 & 0 are members of VLAN
+    "VLAN_BC"    : 0x11, # Ports 4 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 100})    
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x03, # Ports 1 & 0 are members of VLAN
+    "VLAN_BC"    : 0x03, # Ports 1 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 110})
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x03, # Ports 1 & 0 are members of VLAN
+    "VLAN_BC"    : 0x03, # Ports 1 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 111})
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x03, # Ports 1 & 0 are members of VLAN
+    "VLAN_BC"    : 0x03, # Ports 1 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 112})
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x03, # Ports 1 & 0 are members of VLAN
+    "VLAN_BC"    : 0x03, # Ports 1 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 113})
+
 #############################################################################
 # L2 Lookup Parameters Table
 #############################################################################
diff --git a/tools/firmware_generation/SJA1105Q_firmware_sw0_evb.py b/tools/firmware_generation/SJA1105Q_firmware_sw0_evb.py
index e0fb105..fb2df6a 100755
--- a/tools/firmware_generation/SJA1105Q_firmware_sw0_evb.py
+++ b/tools/firmware_generation/SJA1105Q_firmware_sw0_evb.py
@@ -201,6 +201,43 @@ vlan_lookup_table.append({
     "TAG_PORT"   : 0x00, # Egress frames are untagged
     "VLANID"     : default_vlan})
 
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x11, # Ports 4 & 0 are members of VLAN
+    "VLAN_BC"    : 0x11, # Ports 4 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 100})
+
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x09, # Ports 3 & 0 are members of VLAN
+    "VLAN_BC"    : 0x09, # Ports 3 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 110})
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x09, # Ports 3 & 0 are members of VLAN
+    "VLAN_BC"    : 0x09, # Ports 3 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 111})
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x03, # Ports 1 & 0 are members of VLAN
+    "VLAN_BC"    : 0x03, # Ports 1 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 112})
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x05, # Ports 2 & 0 are members of VLAN
+    "VLAN_BC"    : 0x05, # Ports 2 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 113})
+
 #############################################################################
 # L2 Lookup Parameters Table
 #############################################################################
diff --git a/tools/firmware_generation/SJA1105Q_firmware_sw1_bbmini.py b/tools/firmware_generation/SJA1105Q_firmware_sw1_bbmini.py
index 15a2c87..d421081 100755
--- a/tools/firmware_generation/SJA1105Q_firmware_sw1_bbmini.py
+++ b/tools/firmware_generation/SJA1105Q_firmware_sw1_bbmini.py
@@ -201,6 +201,36 @@ vlan_lookup_table.append({
     "TAG_PORT"   : 0x00, # Egress frames are untagged
     "VLANID"     : default_vlan})
 
+# Camera VLANs
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x03, # Ports 1 & 0 are members of VLAN
+    "VLAN_BC"    : 0x03, # Ports 1 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 110})
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x05, # Ports 2 & 0 are members of VLAN
+    "VLAN_BC"    : 0x05, # Ports 2 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 111})
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x09, # Ports 3 & 0 are members of VLAN
+    "VLAN_BC"    : 0x09, # Ports 3 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 112})
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x11, # Ports 4 & 0 are members of VLAN
+    "VLAN_BC"    : 0x11, # Ports 4 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 113})
+
 #############################################################################
 # L2 Lookup Parameters Table
 #############################################################################
diff --git a/tools/firmware_generation/SJA1105Q_firmware_sw1_evb.py b/tools/firmware_generation/SJA1105Q_firmware_sw1_evb.py
index 5876458..5c6b714 100755
--- a/tools/firmware_generation/SJA1105Q_firmware_sw1_evb.py
+++ b/tools/firmware_generation/SJA1105Q_firmware_sw1_evb.py
@@ -201,6 +201,22 @@ vlan_lookup_table.append({
     "TAG_PORT"   : 0x00, # Egress frames are untagged
     "VLANID"     : default_vlan})
 
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x03, # Ports 1 & 0 are members of VLAN
+    "VLAN_BC"    : 0x03, # Ports 1 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 110})
+
+vlan_lookup_table.append({
+    "VING_MIRR"  : 0,
+    "VEGR_MIRR"  : 0,
+    "VMEMB_PORT" : 0x05, # Ports 2 & 0 are members of VLAN
+    "VLAN_BC"    : 0x05, # Ports 2 & 0 are in broadcast domain
+    "TAG_PORT"   : 0x01, # Egress frames are tagged on port 0
+    "VLANID"     : 111})
+
 #############################################################################
 # L2 Lookup Parameters Table
 #############################################################################
-- 
2.11.0

